### DISCUSSION TITLE
scvelo plot for list of variables.

### DISCUSSION THREAD

**Discussion by coolmak32** [theislab/scvelo#465] [Q&A] [answered]
Hi! I have been figuring out the usage of scvelo on single cell Rna seq 10x data. I am a molecular biologist and not that familiar with python coding. So, far I have been able to generate basic graph for Rna velocity using default commands as presented in API section. I have a few queries which I need help with. 

1) I would like to extract top100 genes velocity genes, so I give the following commands.
scv.tl.rank_velocity_genes(adata, groupby='clusters')

KeyError: 'clusters'  ----> I am assuming that this error is because the term "clusters" is not defined. I do not know how can I define it.

so next I just changed str groupby = None.

So I went ahead to generate scatter plot.

scv.pl.scatter(adata, basis=adata.uns['rank_velocity_genes']['names']['Beta'][:3])

ValueError: no field of name Beta. -------> what do I put here? Also, in the "names", I put gene short name but then I get a key error.

I went ahead to try to see if I can extract the data frame

pd.DataFrame(adata.uns['rank_velocity_genes']['names']).head()

I received an error that pd is not defined
so I gave the following command : import pandas as pd. and I give the pd.DataFrame(adata.uns['rank_velocity_genes']['Krt19']).head() again with a gene name.

error: Keyerror: "Krt19".

I am sort of lost as how to define these different key errors.

**Comment by WeilerP** [comment, marked as answer]
@coolmak32, sorry for the late reply.

> 1. I would like to extract top100 genes velocity genes, so I give the following commands.
>    scv.tl.rank_velocity_genes(adata, groupby='clusters')
> 
> KeyError: 'clusters' ----> I am assuming that this error is because the term "clusters" is not defined. I do not know how can I define it.

You need to provide a valid column name of `adata.uns`.  Based on your brief description, I think what you want to do is

```python
top_genes = adata.var['fit_likelihood'].sort_values(ascending=False).index[:100]
```

`rank_velocity_genes` does not return the top 100 velocity genes (based on likelihood) but performs a differential expression (DE) test based on the velocity estimates. Please have a look at our [tutorials](https://scvelo.readthedocs.io/getting_started/) (the suggested code is used e.g. [here](https://scvelo.readthedocs.io/Pancreas/)). Please also have a look at out documentation (e.g. [here](https://scvelo.readthedocs.io/scvelo.tl.rank_velocity_genes/#scvelo.tl.rank_velocity_genes) for `rank_velocity_genes`).
 
> so next I just changed str groupby = None.

By default, this will run the DE test on `velocity_clusters`.

> scv.pl.scatter(adata, basis=adata.uns['rank_velocity_genes']['names']['Beta'][:3])
> 
> ValueError: no field of name Beta. -------> what do I put here? Also, in the "names", I put gene short name but then I get a key error.

What are you trying to achieve here? `adata.uns['rank_velocity_genes']['names']` is not a dictionary (or DataFrame). It is a `numpy.recarray` (see [here](https://numpy.org/doc/stable/reference/generated/numpy.recarray.html)).

> I went ahead to try to see if I can extract the data frame
> 
> pd.DataFrame(adata.uns['rank_velocity_genes']['names']).head()
> 
> I received an error that pd is not defined
> so I gave the following command : import pandas as pd. and I give the pd.DataFrame(adata.uns['rank_velocity_genes']['Krt19']).head() again with a gene name.
> 
> error: Keyerror: "Krt19".

Again: What are you trying to achieve? `'Krt19'` is not a key of `adata.uns['rank_velocity_genes']`. To store the DE ranked genes per clusters in a DataFrame, you can run

```python
scv.DataFrame(adata.uns['rank_velocity_genes']['names'])
```
